### ISSUE TITLE
Error when updating a service backend: missing required field 'ServiceVersion'

### ISSUE THREAD

**Issue by cdodd** [pulumi/pulumi-fastly#119]
I'm getting an issue when I attempt to update a backend of an existing service. When updating the backend I get the error `missing required field 'ServiceVersion'`. However, if I run `pulumi up` a second time it appears to update the service, but doesn't, I can see the service is still on version 1 in the Fastly UI. If I run `pulumi up` for a third time I get the error again, then the cycle continues.

## Steps to reproduce

I've created an example project here https://github.com/cdodd/fastly-issue

1. Run `pulumi up` and create the service
2. Update the backend origin name to something else (e.g. `"origin2.myservice.com"`)
3. Run `pulumi up`, apply the change, you should receive the error `missing required field 'ServiceVersion'`
4. Run `pulumi up`, apply the change, change will show complete.
5. Verify service is not updated in Fastly UI
6. Run `pulumi up`, get the error message again
7.  Cycle continues

### Example output on failure

```
$ pulumi up

Please choose a stack, or create a new one: dev
Previewing update (dev):
     Type                       Name              Plan       Info
     pulumi:pulumi:Stack        fastly-issue-dev
 ~   └─ fastly:index:Servicev1  my-service        update     [diff: ~backends]

Resources:
    ~ 1 to update
    1 unchanged

Do you want to perform this update? yes
Updating (dev):
     Type                       Name              Status                  Info
     pulumi:pulumi:Stack        fastly-issue-dev  **failed**              1 error
 ~   └─ fastly:index:Servicev1  my-service        **updating failed**     [diff: ~backends]; 1 err

Diagnostics:
  pulumi:pulumi:Stack (fastly-issue-dev):
    error: update failed

  fastly:index:Servicev1 (my-service):
    error: 1 error occurred:
    	* updating urn:pulumi:dev::fastly-issue::fastly:index/servicev1:Servicev1::my-service: 1 error occurred:
    	* missing required field 'ServiceVersion'

Resources:
    1 unchanged

Duration: 1s
```

### Example output on update that appears to work, but fails

```
$ pulumi up

Please choose a stack, or create a new one: dev
Previewing update (dev):
     Type                       Name              Plan       Info
     pulumi:pulumi:Stack        fastly-issue-dev
 ~   └─ fastly:index:Servicev1  my-service        update     [diff: ~backends]

Resources:
    ~ 1 to update
    1 unchanged

Do you want to perform this update? yes
Updating (dev):
     Type                       Name              Status      Info
     pulumi:pulumi:Stack        fastly-issue-dev
 ~   └─ fastly:index:Servicev1  my-service        updated     [diff: ~backends]

Resources:
    ~ 1 updated
    1 unchanged

Duration: 18s
```

**Comment by leezen** [comment]
@cdodd Do you mind providing what the state looks like (`pulumi stack export > export.json`) for the `Servicev1` resource? I'm curious what `clonedVersion` is set to.

**Comment by farvour** [comment]
This seems like it is due to the following Terraform change:
https://github.com/fastly/terraform-provider-fastly/commit/a89320487690b29c17bb0ea113b6929cf36691c5

It appears that Pulumi is never going to work correctly with this since the logic in the underlying provider added a an OR condition if the value is not set and I'm unable to actually pass it in as a parameter (nor would I want to).

I had to roll back to an older provider (3.4.x) to work with the pulumi provider again.

**Comment by StarpTech** [comment]
Hi, I run into the same issue. It's unusable.

**Comment by leezen** [comment]
My understanding from looking at that change is that the only scenario in which you're stuck is if you're using a version of the provider prior to `3.0` since that's when the change mentioned above gets picked up. If you're upgrading from a previous version where the state didn't include `ServiceVersion` then you should be able to continue via `pulumi up --refresh` to pick up the correct value and move forward. Does that work for you?

**Comment by StarpTech** [comment]
> My understanding from looking at that change is that the only scenario in which you're stuck is if you're using a version of the provider prior to 3.0

I run into that issue with the latest version. I started a few days ago with the development. Did you test it?

**Comment by StarpTech** [comment]
@leezen doesn't work

```
Diagnostics:
  fastly:index:ServiceCompute (service):
    error: 1 error occurred:
        * updating urn:pulumi:dev::service::fastly:index/serviceCompute:ServiceCompute::service: 1 error occurred:
        * missing required field 'ServiceVersion'

```

**Comment by leezen** [comment]
Apologies -- I thought that `refresh` should bring this value into state, but I was wrong and it does not. Even with using legacy plugin search.  A few experiments:

1) Create a `Servicev1` from scratch at fastly version 3.10.0 -- this is successful.
1) Make a modification and run an update. This fails w/ the `ServiceVersion` issue mentioned above.
3) Change a value and run an update. This does not work.
2) Create a `Servicev1` from scratch at fastly version 3.4.0 -- this is successful. 
3) Change a value and run an update. This works.
3) Upgrade to 3.10.0, make a modification, and run an update. This fails w/ the `ServiceVersion` issue mentioned above.

Looking at the state, it doesn't appear we ever actually write `activeVersion` or `clonedVersion` to state in the newer version of the provider (whereas in 3.4.0, it's present in the state)

**Comment by leezen** [comment]
I think the issue is actually introduced via https://github.com/fastly/terraform-provider-fastly/pull/388

**Comment by farvour** [comment]
@leezen thank you for digging further into this. Do you know if this is something that Pulumi's tf2pulumi has to account for or is this something the upstream provider is going to have to be patched to compensate for? It *seems* more of a pulumi issue as you describe, the value never gets written to state even with a refresh and I think that logic is independent of the upstream provider since the pulumi generator creates a bridge library if I recall...

What is the recommended course-of-action to ensure those of us who were using the provider pre-3.10.0 so that we can upgrade to 3.10.0? It appears that 3.10.0 simply doesn't work at all with Pulumi at this point since you can't even run an update against a freshly created service w/ the 3.10.0 provider...

**Comment by oatnog** [comment]
I'm also running into this. It's my first time using Pulumi.

**Comment by shaidar** [comment]
Tried downgrading to `3.7.0` and then to `3.4.0` and in both cases received a not so descriptive error:

```error: unknown resource type '<nil>'```

So currently am blocked on being able to use this provider.

**Comment by jkodroff** [comment]
@shaidar Can you provide some sample code? I believe the code that the OP provided is pre-V1. (I don't see `ServiceV1` in [the provider docs](https://www.pulumi.com/registry/packages/fastly/api-docs/) anymore.)

**Comment by shaidar** [comment]
Tested out new version of `pulumi-fastly` provider (4.0.1) and ran into the same issue. Ran pulumi up with the minimal example below. It created the Fastly distro successfully, however after tweaking the code by changing the domain comment, it threw the infamous `ServiceVersion error`

```
# Original minimal example

import pulumi
import pulumi_fastly as fastly

ol_devops_qa = fastly.ServiceVcl("ol-devops-qa",
    backends=[fastly.ServiceVclBackendArgs(
        address="127.0.0.1",
        name="localhost",
        port=80,
    )],
    domains=[fastly.ServiceVclDomainArgs(
        comment="ol-devops-qa",
        name="ol-devops-qa.odl.mit.edu",
    )],
    force_destroy=True)
```

```
# Tweaked minimal example by changing comment

import pulumi
import pulumi_fastly as fastly

ol_devops_qa = fastly.ServiceVcl("ol-devops-qa",
    backends=[fastly.ServiceVclBackendArgs(
        address="127.0.0.1",
        name="localhost",
        port=80,
    )],
    domains=[fastly.ServiceVclDomainArgs(
        comment="ol-devops-test",
        name="ol-devops-qa.odl.mit.edu",
    )],
    force_destroy=True)
```

**Comment by shaidar** [comment]
Just checking in to see whether someone had a chance to come up with a potential solution to this?

**Comment by blarghmatey** [comment]
So, I've been digging through the Terraform provider code and the schema.json for this plugin to try and understand what is happening. The overall problem is that the `ServiceVersion` parameter that the TF provider and Fastly library are relying on is an implicit attribute that the TF provider computes automatically. My assumption is that the missing piece is that the Pulumi provider needs to access that attribute in the generated state from the TF provider and include it in its own state management to pass back and forth.

Because it's not an _explicit_ attribute that is managed through the schema.json I'm not sure if that's possible/straightforward? Perhaps we just need to add the `ServiceVersion` to the schema.json to see if we can pass it through explicitly to the TF provider? Or is there some other configuration that we can specify to control which state values are saved from the TF provider output?

**Comment by jkodroff** [comment]
@blarghmatey This should be fixed in pulumi-fastly v4.0.3.

**Comment by kevinrobbins** [comment]
This appears to be fixed for me in v4.0.3!

**Comment by stack72** [comment]
Thanks for taking the time to see this through @jkodroff
